rs/registry/node_provider_rewards/src/v1_rewards.rs

```diff
@@ -0,0 +1,692 @@
+        if rewardable_nodes_count < FULL_REWARDS_MACHINES_LIMIT {
+            logger().add_entry(LogEntry::NodeCountRewardables {
+                node_type: node_type.clone(),
+                region: region.clone(),
+                count: *node_count as usize,
+            });
+
+            rewards_xdr_total.push(rewards_xdr_no_penalty);
+        } else {
+            let mut rewards_multipliers = assigned_multipliers
+                .get(&(region.clone(), node_type.clone()))
+                .cloned()
+                .unwrap_or_default();
+            let assigned_len = rewards_multipliers.len();
+
+            rewards_multipliers.resize(*node_count as usize, unassigned_multiplier);
+
+            logger().add_entry(LogEntry::PerformanceBasedRewardables {
+                node_type: node_type.clone(),
+                region: region.clone(),
+                count: *node_count as usize,
+                assigned_multipliers: rewards_multipliers[..assigned_len].to_vec(),
+                unassigned_multipliers: rewards_multipliers[assigned_len..].to_vec(),
+            });
+
+            for multiplier in rewards_multipliers {
+                rewards_xdr_total.push(Operation::Multiply(*xdr_permyriad, multiplier));
+            }
+        }
+    }
+
+    let rewards_xdr_total = logger().execute(
+        "Compute total permyriad XDR",
+        Operation::SumOps(rewards_xdr_total),
+    );
+    let rewards_xdr_no_reduction_total = logger().execute(
+        "Compute total permyriad XDR no performance penalty",
+        Operation::SumOps(rewards_xdr_no_penalty_total),
+    );
+    logger().add_entry(LogEntry::RewardsXDRTotal(rewards_xdr_total));
+
+    Rewards {
+        xdr_permyriad: rewards_xdr_total.to_u64().unwrap(),
+        xdr_permyriad_no_reduction: rewards_xdr_no_reduction_total.to_u64().unwrap(),
+    }
+}
+
+fn node_providers_rewardables(
+    nodes: &[RewardableNode],
+) -> AHashMap<PrincipalId, RewardablesWithNodesMetrics> {
+    let mut node_provider_rewardables: AHashMap<PrincipalId, RewardablesWithNodesMetrics> =
+        AHashMap::default();
+
+    nodes.iter().for_each(|node| {
+        let (rewardable_nodes, assigned_metrics) = node_provider_rewardables
+            .entry(node.node_provider_id)
+            .or_default();
+
+        let nodes_count = rewardable_nodes
+            .entry((node.region.clone(), node.node_type.clone()))
+            .or_default();
+        *nodes_count += 1;
+
+        if let Some(daily_metrics) = &node.node_metrics {
+            assigned_metrics.insert(node.clone(), daily_metrics.clone());
+        }
+    });
+
+    node_provider_rewardables
+}
+
+pub fn calculate_rewards(
+    days_in_period: u64,
+    rewards_table: &NodeRewardsTable,
+    rewardable_nodes: &[RewardableNode],
+) -> RewardsPerNodeProvider {
+    let mut rewards_per_node_provider = AHashMap::default();
+    let mut rewards_log_per_node_provider = AHashMap::default();
```

Comment:
what's the reason to not just use standard HashMap here?

Reply:
(also, I'm asking b/c adding new packages to registry increases wasm size, which we prefer not to do at present)